source/Cosmos.HAL2/BlockDevice/Ports/Satapi.cs

```diff
@@ -0,0 +1,170 @@
+                return;
+            }
+            mSATAPIDebugger.Send("SATAPI Constructor");
+
+            mPortReg = aSATAPIPort;
+
+            mBlockSize = 2048;
+        }
+
+        public void SendSATAPICommand(ATACommands aCommand, uint aStart, uint aCount)
+        {
+            mPortReg.IS = unchecked((uint)-1);
+
+            int xSlot = FindCMDSlot(mPortReg);
+            if (xSlot == -1) return;
+
+            HBACommandHeader xCMDHeader = new HBACommandHeader(mPortReg.CLB, (uint)xSlot);
+            xCMDHeader.CFL = 5;
+            xCMDHeader.ATAPI = 1;
+            xCMDHeader.PRDTL = (ushort)(((aCount - 1) >> 4) + 1);
+            xCMDHeader.Write = 0;
+
+            xCMDHeader.CTBA = Heap.MemAlloc(128 + ((uint)xCMDHeader.PRDTL) * 16);
+
+            HBACommandTable xCMDTable = new HBACommandTable(xCMDHeader.CTBA, xCMDHeader.PRDTL);
+
+            uint DataBaseAddress = 0x0046C000;
+            for (int i = 0; i < xCMDHeader.PRDTL - 1; i++)
+            {
+                xCMDTable.PRDTEntry[i].DBA = DataBaseAddress;
+                xCMDTable.PRDTEntry[i].DBC = 8 * 1024 - 1;   // 8K bytes (this value should always be set to 1 less than the actual value)
+                xCMDTable.PRDTEntry[i].InterruptOnCompletion = 1;
+                DataBaseAddress += 8 * 1024;    // 4K words
+                aCount -= 16;    // 16 sectors
+            }
+
+            // Last entry
+            xCMDTable.PRDTEntry[xCMDHeader.PRDTL - 1].DBA = DataBaseAddress;
+            xCMDTable.PRDTEntry[xCMDHeader.PRDTL - 1].DBC = aCount * 512 - 1;   // 8K bytes (this value should always be set to 1 less than the actual value)
+            xCMDTable.PRDTEntry[xCMDHeader.PRDTL - 1].InterruptOnCompletion = 1;
+
+            FISRegisterH2D xCMDFIS = new FISRegisterH2D(xCMDTable.CFIS)
+            {
+                FISType = (byte)FISType.FIS_Type_RegisterH2D,
+                IsCommand = 1,
+                Command = (byte)ATACommands.Packet,
+                Device = 1 << 4
+            };
+
+            byte[] xATAPICMD = new byte[12];
+            xATAPICMD[0] = (byte)aCommand;
+            xATAPICMD[2] = (byte)((aStart >> 0x18) & 0xFF);
+            xATAPICMD[3] = (byte)((aStart >> 0x10) & 0xFF);
+            xATAPICMD[4] = (byte)((aStart >> 0x08) & 0xFF);
+            xATAPICMD[5] = (byte)((aStart >> 0x00) & 0xFF);
+            xATAPICMD[9] = (byte)(aCount);
+            for (uint i = 0; i < xATAPICMD.Length; i++)
+            new Core.MemoryBlock(xCMDTable.ACMD, 12).Bytes[i] = xATAPICMD[i];
+            
+            int xSpin = 0;
+            do xSpin++; while ((mPortReg.TFD & 0x88) != 0 && xSpin < 1000000);
+
+            if (xSpin == 1000000)
+            {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.Write("\n[Error]: ");
+                Console.Write("Port timed out!");
+                Console.ResetColor();
+                return;
```

Comment:
This impossible condition or only debug? You should use mSATADebugger.Send() in these cases...